src/types/typeobject.rs

```diff
@@ -177,6 +190,38 @@ impl<'py> PyTypeMethods<'py> for Bound<'py, PyType> {
     {
         self.is_subclass(&T::type_object_bound(self.py()))
     }
+
+    fn mro(&self) -> PyResult<Bound<'py, PyTuple>> {
+        #[cfg(any(Py_LIMITED_API, PyPy))]
+        let mro = self.getattr(intern!(self.py(), "__mro__"))?.extract()?;
+
+        #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+        let mro = unsafe {
+            (*self.as_type_ptr())
+                .tp_mro
+                .assume_borrowed_or_err(self.py())
+        }?
+        .to_owned()
+        .downcast_into()?;
+
+        Ok(mro)
+    }
+
+    fn bases(&self) -> PyResult<Bound<'py, PyTuple>> {
+        #[cfg(any(Py_LIMITED_API, PyPy))]
+        let bases = self.getattr(intern!(self.py(), "__bases__"))?.extract()?;
+
+        #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+        let bases = unsafe {
+            (*self.as_type_ptr())
+                .tp_bases
+                .assume_borrowed_or_err(self.py())
+        }?
+        .to_owned()
+        .downcast_into()?;
```

Comment:
Is it necessary to do a safe downcast? As I understand, a tuple is guaranteed here from Python itself. That makes me think we can save the implicit type check and use `downcast_unchecked`

Reply:
I checked the CPython source directly, just for the sake of it, and yeah, can confirm this must be a tuple. Both have `PyTuple_CheckExact` guards on them:

```c
static inline void
set_tp_bases(PyTypeObject *self, PyObject *bases)
{
    assert(PyTuple_CheckExact(bases));
    if (self->tp_flags & _Py_TPFLAGS_STATIC_BUILTIN) {
        // XXX tp_bases can probably be statically allocated for each
        // static builtin type.
        assert(_Py_IsMainInterpreter(_PyInterpreterState_GET()));
        assert(self->tp_bases == NULL);
        if (PyTuple_GET_SIZE(bases) == 0) {
            assert(self->tp_base == NULL);
        }
        else {
            assert(PyTuple_GET_SIZE(bases) == 1);
            assert(PyTuple_GET_ITEM(bases, 0) == (PyObject *)self->tp_base);
            assert(self->tp_base->tp_flags & _Py_TPFLAGS_STATIC_BUILTIN);
            assert(_Py_IsImmortal(self->tp_base));
        }
        _Py_SetImmortal(bases);
    }
    self->tp_bases = bases;
}
```


```c
static inline void
set_tp_mro(PyTypeObject *self, PyObject *mro)
{
    assert(PyTuple_CheckExact(mro));
    if (self->tp_flags & _Py_TPFLAGS_STATIC_BUILTIN) {
        // XXX tp_mro can probably be statically allocated for each
        // static builtin type.
        assert(_Py_IsMainInterpreter(_PyInterpreterState_GET()));
        assert(self->tp_mro == NULL);
        /* Other checks are done via set_tp_bases. */
        _Py_SetImmortal(mro);
    }
    self->tp_mro = mro;
}```